components/eam/src/physics/cam/micro_p3.F90

```diff
@@ -2855,6 +2868,125 @@ subroutine prevent_ice_overdepletion(pres,t_atm,qv,latent_heat_sublim,inv_dt,
 
 end subroutine prevent_ice_overdepletion
 
+subroutine water_vapor_conservation(qv,qidep,qinuc,qi2qv_sublim_tend,qr2qv_evap_tend,dt)
+
+  implicit none
+
+  real(rtype), intent(in)     :: qv,qi2qv_sublim_tend,qr2qv_evap_tend,dt
+  real(rtype), intent(inout) :: qidep,qinuc
+  real(rtype)                :: qv_avail, qv_sink, ratio
+
+  qv_avail = qv + (qi2qv_sublim_tend+qr2qv_evap_tend)*dt
+  qv_sink  = (qidep + qinuc)*dt
+
+  if (qv_sink > qv_avail .and. qv_sink>1.e-20_rtype) then
+     ratio = qv_avail/qv_sink
+     qidep = qidep*ratio
+     qinuc = qinuc*ratio     
+  endif
+
+  return
+end subroutine water_vapor_conservation
+
+subroutine ice_supersat_conservation(qidep,qinuc,cld_frac_i,qv,qv_sat_i,latent_heat_sublim,T_atm,dt)
+  !Make sure ice processes don't drag qv below ice supersaturation
+  
+  implicit none
+
+  real(rtype), intent(in) :: cld_frac_i,qv,qv_sat_i,latent_heat_sublim,T_atm,dt
+  real(rtype), intent(inout) :: qidep,qinuc
+  
+  real(rtype) :: qv_sink, qv_avail, fract 
+
+  qv_sink = qidep + qinuc ! in [kg/kg] cell-avg values
+
+  if (qv_sink > qsmall .and. cld_frac_i > 1.0e-20_rtype) then
+     ! --- Available water vapor for deposition/nucleation
+     qv_avail = (qv - qv_sat_i) / &
```

Comment:
@JS-WRF-SBM - I think qv_avail should be end-of-step qv, so the line above should be "qv_avail = (qv +(qi2qv_sublim_tend+qr2qv_evap_tend)*dt -qv_sat_i)/(stuff that was there before, or maybe abi). Do you agree? Further, I think that ice_supersat_conservation subroutine is always more strict than water_vapor_conservation so we could probably completely delete the latter function. Do you agree?

Reply:
@PeterCaldwell 

I agree that for the check of the available ice supersaturation we should use the 'full potential' of the water vapor, which seems to be at the beginning of the MP time step, rather then at the end of it -- isn't it?

I'm really concerned  from all these non-bounded evaporative processes (ice sublim / rain evap), particulary for LR -- this is why I tend to keep the check.

Reply:
@JS-WRF-SBM - I think I rewrote my comment while you were responding to it. Maybe the new version answers your questions, but to be clear:

1. In terms of the 'full potential' of the water vapor, I think an easier and equivalent way to think is "at the end of the timestep (i.e. after this function and the state update), the smallest ice supersaturation should be is 0%". Your current version rescales such that "in the absence of qv sources, supersaturation will be zero at the end of the step". But there are qv sources so your rescaling will result in positive supersaturation at the end of the timestep... it is too aggressive.

2. I'm arguing that if you ran your code with and without qv_conservation, you'd get BFB answers... qv_conservation just wastes flops and makes the code more complicated. This is something we should test empirically by doing runs with and without this fixer. I'm not sure when I'll find time to do that though.

Reply:
@PeterCaldwell 

-  The deposition and sublimation (sublimation being a source of `qv`) occurs alternately per grid point. So I wouldn't worry about that. If sublimation is enabled, both `qinuc` and `qidep` needs to be zero. Did we check that this is the case for `qinuc` which comes from outside the P3? I personally haven't checked that.
-  The rain evaporation requires `qv - qv_sat_l < 0.0` 
-   Not sure I'm fully following your argument above - but if you argue that limiting only the qv > 0 at the `update_liquid/ice` routines is enough? Please note that when you involve `qv_sat_i`  / `qv_sat_l` you have an intrinsic dependency on temperature from processes before P3. This is a more general condition compare to `qv > 0`

Reply:
@JS-WRF-SBM 
- good point that if you are sublimating or evaporating you shouldn't be doing vapor deposition or ice nucleation. There is a check in p3 that only gives qinuc positive values if ice supersaturation is > 5%. I concede that this means that your original implementation where qinuc and qidep are limited based on qv rather than qv+(sublim+evap)*dt is sufficient. I think this means you could have also omitted sublim + evap from your qv_limit function. I'm not going to make changes to what I've done because it isn't wrong (just adds something we know will be zero) and the logic for why it is ok to leave off these source terms is a bit convoluted. But mostly because I don't want to change the PR. 
- I'm confused about your last bullet. My argument here is basically that if you insist that qv is at least saturated with respect to ice, then you're already ensuring that qv>0 because qv_sat_i is always > 0. Thus there's no point in your qv_limiter function. It is true that abi has a mild dependence on T. I'm not sure where you're getting the qv_sat_l dependence - from rain evaporation? I agree that would have an impact, but I think it is in the noise compared to other assumptions we make.